### ISSUE TITLE
[Security] Possibility of data leakage between projects

### ISSUE THREAD

**Issue by nicoinn** [hopshadoop/hops#317]
%spark.pyspark allows to write files to project__user HDFS home that are readable from inside other projects allowing for data leakage between projects. 

Guide to reproduce
-------------------

In project "IP2Location", I create a dataset "IP2location_raw" and put a file IP2LOCATION-LITE-DB11.CSV". Then I create a Zeppelin notebook and do: 

```
%spark.pyspark
from pyspark.sql import SQLContext
sqlContext = SQLContext(sc)

df = sqlContext.read.format('com.databricks.spark.csv').options(header='false', inferschema='true', delimiter=",").load('hdfs://10.200.1.100:8020/Projects/IP2location/IP2location_raw/IP2LOCATION-LITE-DB11.CSV')

df.write.mode("overwrite").parquet("ip2location.parquet")
```

Create a new project "Testing" and create a new notebook. 
Running

```
%spark.pyspark
from pyspark.sql import SQLContext
sqlContext = SQLContext(sc)

ip2loc = sqlContext.read.parquet("/user/IP2location/ip2location.parquet/")
ip2loc.show()
```
 
will successfully read the file from the other project causing a data leak!

Note that 
````
df = sqlContext.read.format('com.databricks.spark.csv').options(header='false', inferschema='true', delimiter=",").load('hdfs://10.200.1.100:8020/Projects/IP2location/IP2location_raw/IP2LOCATION-LITE-DB11.CSV') ```

within the Testing project fails with a permission denied as it should be.

**Comment by nicoinn** [comment]
And a little bit more debugging info:

A `hadoop fs -ls` from the command line gives

```
drwxr-xr-x   - IP2location hadoop          0 2017-01-29 12:13 /user/IP2location/ip2location.parquet
```

which should probably have permissions drwx------ 

--------

Trying to write back to a dataset from pyspark, I get a permission deny

```
%spark.pyspark
df.write.mode("overwrite").parquet("hdfs://10.200.1.100:8020/Projects/IP2location/IP2location_raw/ip2location.parquet")
```

```
Traceback (most recent call last):
  File "/tmp/zeppelin_pyspark-2305378312749931362.py", line 267, in <module>
    raise Exception(traceback.format_exc())
Exception: Traceback (most recent call last):
  File "/tmp/zeppelin_pyspark-2305378312749931362.py", line 265, in <module>
    exec(code)
  File "<stdin>", line 1, in <module>
  File "/srv/spark/python/pyspark/sql/readwriter.py", line 641, in parquet
    self._jwrite.parquet(path)
  File "/srv/spark/python/lib/py4j-0.10.3-src.zip/py4j/java_gateway.py", line 1133, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/srv/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/srv/spark/python/lib/py4j-0.10.3-src.zip/py4j/protocol.py", line 319, in get_return_value
    format(target_id, ".", name), value)
Py4JJavaError: An error occurred while calling o242.parquet.
: org.apache.hadoop.security.AccessControlException: Permission denied: user=IP2location, access=WRITE, inode=470349
	at org.apache.hadoop.hdfs.server.namenode.FSPermissionChecker.check(FSPermissionChecker.java:264)
	at org.apache.hadoop.hdfs.server.namenode.FSPermissionChecker.check(FSPermissionChecker.java:236)
	at org.apache.hadoop.hdfs.server.namenode.FSPermissionChecker.check(FSPermissionChecker.java:224)
	at org.apache.hadoop.hdfs.server.namenode.FSPermissionChecker.checkPermission(FSPermissionChecker.java:160)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.checkPermission(FSNamesystem.java:4584)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.checkAncestorAccess(FSNamesystem.java:4558)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.mkdirsInternal(FSNamesystem.java:2981)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.mkdirsInt(FSNamesystem.java:2949)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.access$3300(FSNamesystem.java:206)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem$25.performTask(FSNamesystem.java:2929)
	at io.hops.transaction.handler.TransactionalRequestHandler.execute(TransactionalRequestHandler.java:112)
	at io.hops.transaction.handler.HopsTransactionalRequestHandler.execute(HopsTransactionalRequestHandler.java:50)
	at io.hops.transaction.handler.RequestHandler.handle(RequestHandler.java:54)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.mkdirs(FSNamesystem.java:2936)
	at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.mkdirs(NameNodeRpcServer.java:636)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.mkdirs(ClientNamenodeProtocolServerSideTranslatorPB.java:550)
	at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java:61847)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:616)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:983)
	at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2049)
	at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2045)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1712)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2045)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.apache.hadoop.ipc.RemoteException.instantiateException(RemoteException.java:106)
	at org.apache.hadoop.ipc.RemoteException.unwrapRemoteException(RemoteException.java:73)
	at org.apache.hadoop.hdfs.DFSClient.primitiveMkdir(DFSClient.java:2546)
	at org.apache.hadoop.hdfs.DFSClient.mkdirs(DFSClient.java:2505)
	at org.apache.hadoop.hdfs.DistributedFileSystem.mkdirs(DistributedFileSystem.java:722)
	at org.apache.hadoop.fs.FileSystem.mkdirs(FileSystem.java:1881)
	at org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter.setupJob(FileOutputCommitter.java:313)
	at org.apache.spark.sql.execution.datasources.BaseWriterContainer.driverSideSetup(WriterContainer.scala:110)
	at org.apache.spark.sql.execution.datasources.InsertIntoHadoopFsRelationCommand$$anonfun$run$1.apply$mcV$sp(InsertIntoHadoopFsRelationCommand.scala:140)
	at org.apache.spark.sql.execution.datasources.InsertIntoHadoopFsRelationCommand$$anonfun$run$1.apply(InsertIntoHadoopFsRelationCommand.scala:115)
	at org.apache.spark.sql.execution.datasources.InsertIntoHadoopFsRelationCommand$$anonfun$run$1.apply(InsertIntoHadoopFsRelationCommand.scala:115)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:57)
	at org.apache.spark.sql.execution.datasources.InsertIntoHadoopFsRelationCommand.run(InsertIntoHadoopFsRelationCommand.scala:115)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:60)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:58)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.doExecute(commands.scala:74)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:115)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:115)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:136)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:133)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:114)
	at org.apache.spark.sql.execution.QueryExecution.toRdd$lzycompute(QueryExecution.scala:86)
	at org.apache.spark.sql.execution.QueryExecution.toRdd(QueryExecution.scala:86)
	at org.apache.spark.sql.execution.datasources.DataSource.write(DataSource.scala:510)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:211)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:194)
	at org.apache.spark.sql.DataFrameWriter.parquet(DataFrameWriter.scala:488)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:237)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:280)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:214)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.security.AccessControlException): Permission denied: user=IP2location, access=WRITE, inode=470349
	at org.apache.hadoop.hdfs.server.namenode.FSPermissionChecker.check(FSPermissionChecker.java:264)
	at org.apache.hadoop.hdfs.server.namenode.FSPermissionChecker.check(FSPermissionChecker.java:236)
	at org.apache.hadoop.hdfs.server.namenode.FSPermissionChecker.check(FSPermissionChecker.java:224)
	at org.apache.hadoop.hdfs.server.namenode.FSPermissionChecker.checkPermission(FSPermissionChecker.java:160)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.checkPermission(FSNamesystem.java:4584)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.checkAncestorAccess(FSNamesystem.java:4558)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.mkdirsInternal(FSNamesystem.java:2981)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.mkdirsInt(FSNamesystem.java:2949)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.access$3300(FSNamesystem.java:206)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem$25.performTask(FSNamesystem.java:2929)
	at io.hops.transaction.handler.TransactionalRequestHandler.execute(TransactionalRequestHandler.java:112)
	at io.hops.transaction.handler.HopsTransactionalRequestHandler.execute(HopsTransactionalRequestHandler.java:50)
	at io.hops.transaction.handler.RequestHandler.handle(RequestHandler.java:54)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.mkdirs(FSNamesystem.java:2936)
	at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.mkdirs(NameNodeRpcServer.java:636)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.mkdirs(ClientNamenodeProtocolServerSideTranslatorPB.java:550)
	at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java:61847)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:616)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:983)
	at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2049)
	at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2045)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1712)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2045)
	at org.apache.hadoop.ipc.Client.call(Client.java:1475)
	at org.apache.hadoop.ipc.Client.call(Client.java:1412)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Invoker.invoke(ProtobufRpcEngine.java:229)
	at com.sun.proxy.$Proxy11.mkdirs(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.invokeMethod(RetryInvocationHandler.java:191)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.invoke(RetryInvocationHandler.java:102)
	at com.sun.proxy.$Proxy11.mkdirs(Unknown Source)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolTranslatorPB.mkdirs(ClientNamenodeProtocolTranslatorPB.java:516)
	at org.apache.hadoop.hdfs.DFSClient$36.doAction(DFSClient.java:2541)
	at org.apache.hadoop.hdfs.DFSClient.doClientActionWithRetry(DFSClient.java:2812)
	at org.apache.hadoop.hdfs.DFSClient.doClientActionWithRetry(DFSClient.java:2764)
	at org.apache.hadoop.hdfs.DFSClient.primitiveMkdir(DFSClient.java:2544)
	... 36 more
```

i.e. Pyspark tries to write the file as user "IP2location", while the folder permissions look like this: 

```drwxr-x---   - IP2location__nicolas_ IP2location__IP2location_raw          0 2017-01-29 12:26 /Projects/IP2location/IP2location_raw ```

------------------------

Finally, repeating the same with %livy.pyspark instead of %spark.pyspark: 
- No permission deny while writing to dataset
- ``df.write.mode("overwrite").parquet("ip2location.parquet")`` puts the file into ``/user/IP2location__nicolas_/ip2location.parquet`` instead of ``/user/IP2location/ip2location.parquet/`` 
- Leaking data is still possible


To summarise: %spark.pyspark runs as "project" and %livy.pyspark runs as "project__user". Both write files with permissions drwxr-xr-x to the home folder.

**Comment by jimdowling** [comment]
I think i have found the issue here. The spark staging directory is created by default in /user/<projectname>/.spark_staging
The user should absolutely not be able to read/write to '/user' directories. We will change the spark staging directory in the job launcher/zeppelin so that it is set to a directory local to the project.
Thanks @nicoinn
